### PR TITLE
Sams changes to John's share-edges branch

### PR DESCRIPTION
**Pull request by smason**
proposed tasks were listed as:

- [x] Automated test - for already fixed bit
- [ ] Test for known bug (vertex/node confusion) - with a curved surface
- [x] Test and bugfix for faceting tolerance in hdf5 output
- [x] MOAB error handling

Then, when we have it working properly and with some good automated tests:
- [ ] Refactor - e.g. replace vertex inserter, remove one level of abstraction, tidy up structure, move to overlap_checker.
- [ ] Add a tool to check that we don't need a better vertex inserter.

### CODE REVIEW

**Comment by smason** [comment]
@johnnonweiler when you're back!

re: vertex/node confusion.  AFAICT it's doing the right thing for me. I've made a torus and have had a look at the resulting h5m file and everything looks okay in my viewer and check_watertight says all okay

File: src/MBTool.cpp [hunk 1]
```diff
@@ -11,6 +11,14 @@ const char geom_categories[][CATEGORY_TAG_SIZE] = {"Vertex\0",
 						   "Group\0"};
 
 
+
+static inline void check_moab_rval(moab::ErrorCode ret, const char * what) {
+  if (ret == moab::MB_SUCCESS) {
+    return;
+  }
+  throw mberror(ret, what);
+}
```
Comment (johnnonweiler) [inline src/MBTool.cpp]: What is the reason for doing it that way round, with the early return, rather than the shorter version with no early return?
```suggestion
static inline void check_moab_rval(moab::ErrorCode ret, const char * what) {
  if (ret != moab::MB_SUCCESS) {
    throw mberror(ret, what);
  }
}
```

File: src/MBTool.cpp [hunk 2]
```diff
@@ -63,64 +71,63 @@ MBTool::MBTool() {
     MB_CHK_SET_ERR_RET(rval, "Error creating mat_id_tag");
 }
 
-// destructor
+// destructore
```
Comment (johnnonweiler) [inline src/MBTool.cpp]: ?

File: src/brep_faceter.cc [hunk 3]
```diff
@@ -289,8 +286,8 @@ void brep_faceter(std::string brep_file, std::string json_file,
   read_metadata(json_file, materials_map);
 
   MBTool mbtool;
-  // TODO: Use faceting tolerance value in set_tags and review use of GEOMETRY_RESABS
-  mbtool.set_tags();
+  // TODO: review use of GEOMETRY_RESABS
+  mbtool.set_faceting_tol_tag(facet_tol.tolerance);
```
Comment (johnnonweiler) [inline src/brep_faceter.cc:290]: This includes a change in behaviour, which should ideally be in a separate commit from refactoring.
(Personally, I wouldn't have changed the behaviour, but that's because I don't know how the output is being used.)

File: src/steps2h5m.cc [hunk 4]
```diff
@@ -16,7 +16,7 @@ void steps2h5m(std::string input_file, const FacetingTolerance &facet_tol, std::
   json j = json::parse(std::ifstream(input_file));
 
   MBTool mbtool;
-  mbtool.set_tags();
+  mbtool.set_faceting_tol_tag(facet_tol.tolerance);
```
Comment (johnnonweiler) [inline src/steps2h5m.cc:19]: (As above,) this change includes a change to behaviour - not just refactoring.

File: src/test/test_faceter.cc [hunk 5]
```diff
@@ -57,23 +54,23 @@ TEST_CASE("Faceting BREP and writing to MOAB", "[faceter]") {
   double xyz[3] = {1.0, 1.0, 1.0};
   ret = gqt.point_in_volume(vol1, xyz, result);
   REQUIRE(ret == moab::MB_SUCCESS);
-  REQUIRE(result == 1);
+  CHECK(result == 1);
```
Comment (johnnonweiler) [inline src/test/test_faceter.cc:57]: I would ideally have these changes from REQUIRE to CHECK in a separate commit, but I think it doesn't matter.

File: src/test/test_occ_faceter_with_cubit_output.sh [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+set -euo pipefail
+
+./faceter_test
+
+python dagmc_canon.py cubit_output.h5m > cubit_output.txt
+python dagmc_canon.py test_output.h5m > test_output.txt
+
+exec diff -u cubit_output.txt test_output.txt
```
Comment (johnnonweiler) [inline src/test/test_occ_faceter_with_cubit_output.sh:10]: Just out of curiousity; Why use "exec" here?

**Review by johnnonweiler — APPROVED** [review]
This all looks good.  The new test looks great, and the MOAB error checking should make things much safer (as well as making the code much nicer to read).

I like the fact that there are lots of examples of replacing >=2 lines with 1 line.

Some of the changes, such as position of "{", changing "std::string" to "const std:string &", or "idx++" to "idx += 1" seem like personal preference changes, which I wonder if we should avoid (to make it easier for Andy to review the changes).

I wonder if we should consider any of the additional error handling features provided by MOAB (see moab/ErrorHandler.hpp) ?  I think probably not.

I think it might be better, to simplify the code, not to write a "what" string for each exception, and to use a macro to report the source-file/line-number (/function-name?).  Also, I think we should put the error code in the "what" message.

(For the record, we've also discussed offline how to follow-up on the torus test.  We do think there is a bug (/some bugs) and should compare output from Cubit for the same geometry.)
